roast/hash.go

```diff
@@ -5,22 +5,57 @@ import (
 	"math/big"
 )
 
-// The contextString as defined in section 6.5. FROST(secp256k1, SHA-256) of
-// [FROST] is "FROST-secp256k1-SHA256-v1". Since we do a BIP-340 specialized
-// version, we use "FROST-secp256k1-BIP340-v1" as the contextString.
-var contextString = []byte("FROST-secp256k1-BIP340-v1")
+// Hash interface abstracts out hash functions implementations specific to the
+// ciphersuite used. This is a strategy design pattern allowing to use FROST
+// with different ciphersuites, like secp256k1 or Jubjub curves.
+//
+// [FROST] requires the use of a cryptographically secure hash function,
+// generically written as H. Using H, [FROST] introduces distinct domain-separated
+// hashes, H1, H2, H3, H4, and H5. The details of H1, H2, H3, H4, and H5 vary
+// based on ciphersuite.
+type Hash interface {
+	H1(m []byte) *big.Int
+	H2(m []byte, ms ...[]byte) *big.Int
+}
+
+// Bip340Hash is [BIP340] implementation of [FROST] functions required by the
+// `Hash` interface.
+type Bip340Hash struct {
+}
+
+// H1 is the implementation of H1(m) function from [FROST] implemented in a way
+// compatible to how [BIP340] hashing functions are specified.
+func (b *Bip340Hash) H1(m []byte) *big.Int {
+	// From [FROST], we know the tag should be DST = contextString || "rho".
+	dst := concat(b.contextString(), []byte("rho"))
+	// We use [BIP340]-compatible hashing algorithm and turn the hash into
+	// a scalar, as expected by [FROST] for H1.
+	return b.hashToScalar(dst, m)
+}
 
-// h1 is the implementation of H1(m) function from [FROST] implemented in a way
+// H2 is the implementation of H2(m) function from [FROST] implemented in a way
 // compatible to how [BIP340] hashing functions are specified.
-func h1(m []byte) *big.Int {
-	dst := concat(contextString, []byte("rho"))
-	return hashToScalar(dst, m)
+func (b *Bip340Hash) H2(m []byte, ms ...[]byte) *big.Int {
+	// From [FROST], we know the tag should be DST = contextString || "chal".
+	dst := concat(b.contextString(), []byte("chal"))
+	// We use [BIP340]-compatible hashing algorithm and turn the hash into
+	// a scalar, as expected by [FROST] for H2.
+	return b.hashToScalar(dst, concat(m, ms...))
+}
```

Comment:
Looking at the prototype, the `H2 function` [as implemented in `hash.go`](https://github.com/threshold-network/roast-go/blob/e33d503c97ad970cc4bbf99463e197fc20fde39e/hash.go#L45-L47) is not used at all and the [`BIP340HashChallenge`](https://github.com/threshold-network/roast-go/blob/e33d503c97ad970cc4bbf99463e197fc20fde39e/hash.go#L23C12-L25) is used instead.

BIP-340 uses three hashing functions:
- `hash_BIP0340/nonce`
- `hash_BIP0340/challenge`
- `hash_BIP0340/aux`

FROST(secp256k1, SHA-256) uses five hashing functions:
- `H1` with `DST = contextString || "rho"`
- `H2` with `DST = contextString || "chal"`
- `H3` with `DST = contextString || "nonce"`
- `H4` as `H(contextString || "msg" || m)`
- `H5` as `H(contextString || "com" || m)`

For consistency, I was thinking about using the `FROST-secp256k1-BIP340-v1 || <tag>` everywhere with `<tag> ∈ {"rho", "chal", "nonce", "msg", "com"}`.

`H1`, `H2`, `H3` would turn the hash into scalar using modulo on secp256k1 curve order and return `*big.Int`. `H4` and `H5` would return `[]byte`.

What do you think about it @eth-r?

Reply:
I fixed it in fc77336053858a58d753f7e8616a4bc4e24ad29d.